Error while checking build/O2/fullCI for 90b8fcd5daea4aa09a8e76f6b6c828c616d3f623 at 2024-07-26 09:20:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
/sw/SOURCES/O2/13333-slc8_x86-64/0/Framework/Foundation/src/RuntimeError.cxx:84:3: error: 'strncpy' was not declared in this scope
/sw/SOURCES/O2/13333-slc8_x86-64/0/Framework/Foundation/test/test_RuntimeError.cxx:36:20: error: 'demangled_backtrace_symbols' is not a member of 'o2::framework'
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13333/90b8fcd5daea4aa09a8e76f6b6c828c616d3f623/build_O2_fullCI/pretty.html).

--- a/Framework/Core/src/DataProcessingDevice.cxx
+++ b/Framework/Core/src/DataProcessingDevice.cxx
@@ -431,7 +431,7 @@ void DataProcessingDevice::Init()
       auto& err = error_from_ref(e);
       O2_SIGNPOST_ID_FROM_POINTER(cid, device, &context);
       O2_SIGNPOST_EVENT_EMIT_ERROR(device, cid, "Init", "Exception caught while in Init: %{public}s. Invoking errorCallback.", err.what);
-      demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
+      BacktraceHelpers::demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
       auto& stats = ref.get<DataProcessingStats>();
       stats.updateStats({(int)ProcessingStatsId::EXCEPTION_COUNT, DataProcessingStats::Op::Add, 1});
       InitErrorContext errorContext{ref, e};
@@ -446,7 +446,7 @@ void DataProcessingDevice::Init()
       auto& context = ref.get<DataProcessorContext>();
       O2_SIGNPOST_ID_FROM_POINTER(cid, device, &context);
       O2_SIGNPOST_EVENT_EMIT_ERROR(device, cid, "Init", "Exception caught while in Init: %{public}s. Exiting with 1.", err.what);
-      demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
+      BacktraceHelpers::demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
       auto& stats = ref.get<DataProcessingStats>();
       stats.updateStats({(int)ProcessingStatsId::EXCEPTION_COUNT, DataProcessingStats::Op::Add, 1});
       exit(1);
@@ -1111,7 +1111,7 @@ void DataProcessingDevice::fillContext(DataProcessorContext& context, DeviceCont
       auto& context = ref.get<DataProcessorContext>();
       O2_SIGNPOST_ID_FROM_POINTER(cid, device, &context);
       O2_SIGNPOST_EVENT_EMIT_ERROR(device, cid, "Run", "Exception while running: %{public}s. Invoking callback.", err.what);
-      demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
+      BacktraceHelpers::demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
       auto& stats = ref.get<DataProcessingStats>();
       stats.updateStats({(int)ProcessingStatsId::EXCEPTION_COUNT, DataProcessingStats::Op::Add, 1});
       ErrorContext errorContext{record, ref, e};
@@ -1126,7 +1126,7 @@ void DataProcessingDevice::fillContext(DataProcessorContext& context, DeviceCont
       ServiceRegistryRef ref{serviceRegistry, ServiceRegistry::globalDeviceSalt()};
       auto& context = ref.get<DataProcessorContext>();
       O2_SIGNPOST_ID_FROM_POINTER(cid, device, &context);
-      demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
+      BacktraceHelpers::demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
       auto& stats = ref.get<DataProcessingStats>();
       stats.updateStats({(int)ProcessingStatsId::EXCEPTION_COUNT, DataProcessingStats::Op::Add, 1});
       switch (errorPolicy) {

--- a/Framework/Core/src/runDataProcessing.cxx
+++ b/Framework/Core/src/runDataProcessing.cxx
@@ -11,6 +11,7 @@
 #define BOOST_BIND_GLOBAL_PLACEHOLDERS
 #include <stdexcept>
 #include "Framework/BoostOptionsRetriever.h"
+#include "Framework/BacktraceHelpers.h"
 #include "Framework/CallbacksPolicy.h"
 #include "Framework/ChannelConfigurationPolicy.h"
 #include "Framework/ChannelMatching.h"
@@ -656,7 +657,7 @@ void handle_crash(int sig)
     auto retVal = write(STDERR_FILENO, buffer, strlen(buffer));
     (void)retVal;
   }
-  demangled_backtrace_symbols(array, size, STDERR_FILENO);
+  BacktraceHelpers::demangled_backtrace_symbols(array, size, STDERR_FILENO);
   {
     char const* msg = "Backtrace complete.\n";
     int len = strlen(msg); /* the byte length of the string */
@@ -982,7 +983,7 @@ void doDPLException(RuntimeErrorRef& e, char const* processName)
          " Reason: {}"
          "\n Backtrace follow: \n",
          processName, err.what);
-    demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
+    BacktraceHelpers::demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
   } else {
     LOGP(fatal,
          "Unhandled o2::framework::runtime_error reached the top of main of {}, device shutting down."
@@ -1928,7 +1929,7 @@ int runStateMachine(DataProcessorSpecs const& workflow,
         } catch (o2::framework::RuntimeErrorRef ref) {
           auto& err = o2::framework::error_from_ref(ref);
 #ifdef DPL_ENABLE_BACKTRACE
-          demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
+          BacktraceHelpers::demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
 #endif
           LOGP(error, "invalid workflow in {}: {}", driverInfo.argv[0], err.what);
           return 1;
@@ -2059,7 +2060,7 @@ int runStateMachine(DataProcessorSpecs const& workflow,
           LOGP(error, "unable to merge configurations in {}: {}", driverInfo.argv[0], err.what);
 #ifdef DPL_ENABLE_BACKTRACE
           std::cerr << "\nStacktrace follows:\n\n";
-          demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
+          BacktraceHelpers::demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
 #endif
           return 1;
         }

--- a/Framework/Foundation/CMakeLists.txt
+++ b/Framework/Foundation/CMakeLists.txt
@@ -16,6 +16,7 @@ install(DIRECTORY ${CMAKE_CURRENT_LIST_DIR}/include/Framework
 o2_add_library(FrameworkFoundation
                SOURCES src/RuntimeError.cxx
                src/Signpost.cxx
+               src/BacktraceHelpers.cxx
                TARGETVARNAME targetName
                PUBLIC_LINK_LIBRARIES O2::FrameworkFoundation3rdparty
               )

--- a/Framework/Foundation/include/Framework/BacktraceHelpers.h
+++ b/Framework/Foundation/include/Framework/BacktraceHelpers.h
@@ -1,0 +1,22 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+#ifndef O2_FRAMEWORK_BACKTRACE_HELPERS_H_
+#define O2_FRAMEWORK_BACKTRACE_HELPERS_H_
+
+namespace o2::framework
+{
+struct BacktraceHelpers {
+  static constexpr unsigned int MAX_BACKTRACE_SIZE = 100;
+  static void demangled_backtrace_symbols(void** backtrace, unsigned int total, int fd);
+};
+} // namespace o2::framework
+
+#endif // O2_FRAMEWORK_BACKTRACE_HELPERS_H_

--- a/Framework/Foundation/include/Framework/RuntimeError.h
+++ b/Framework/Foundation/include/Framework/RuntimeError.h
@@ -11,23 +11,22 @@
 #ifndef O2_FRAMEWORK_RUNTIMEERROR_H_
 #define O2_FRAMEWORK_RUNTIMEERROR_H_
 
+#include "BacktraceHelpers.h"
+
 namespace o2::framework
 {
 
 struct RuntimeError {
   static constexpr unsigned int MAX_RUNTIME_ERRORS = 64;
   static constexpr unsigned int MAX_RUNTIME_ERROR_SIZE = 1024;
-  static constexpr unsigned int MAX_BACKTRACE_SIZE = 100;
   char what[MAX_RUNTIME_ERROR_SIZE];
-  void* backtrace[MAX_BACKTRACE_SIZE];
+  void* backtrace[BacktraceHelpers::MAX_BACKTRACE_SIZE];
   int maxBacktrace = 0;
 };
 
 struct RuntimeErrorRef {
   int index = 0;
 };
-
-void demangled_backtrace_symbols(void** backtrace, unsigned int total, int fd);
 
 RuntimeErrorRef runtime_error(const char*);
 RuntimeErrorRef runtime_error_f(const char*, ...);

--- a/Framework/Foundation/src/BacktraceHelpers.cxx
+++ b/Framework/Foundation/src/BacktraceHelpers.cxx
@@ -1,0 +1,125 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+#include "Framework/BacktraceHelpers.h"
+#include <climits>
+#include <cstddef>
+#include <cstdio>
+#include <cstdlib>
+#include <unistd.h>
+#include <cxxabi.h>
+#include <execinfo.h>
+
+namespace o2::framework
+{
+
+void BacktraceHelpers::demangled_backtrace_symbols(void** stackTrace, unsigned int stackDepth, int fd)
+{
+  char** stackStrings;
+  stackStrings = backtrace_symbols(stackTrace, stackDepth);
+  char exe[PATH_MAX];
+  bool hasExe = false;
+  int exeSize = 0;
+
+#if __linux__
+  exeSize = readlink("/proc/self/exe", exe, PATH_MAX);
+  if (exeSize == -1) {
+    dprintf(fd, "Unable to detect exectuable name\n");
+    hasExe = false;
+  } else {
+    dprintf(fd, "Executable is %.*s\n", exeSize, exe);
+    hasExe = true;
+  }
+#endif
+
+  for (size_t i = 1; i < stackDepth; i++) {
+
+    size_t sz = 64000; // 64K ought to be enough for our templates...
+    char* function = static_cast<char*>(malloc(sz));
+    char *begin = nullptr, *end = nullptr;
+    // find the last space and address offset surrounding the mangled name
+#if __APPLE__
+    for (char* j = stackStrings[i]; *j; ++j) {
+      if (*j == ' ' && *(j + 1) != '+') {
+        begin = j;
+      } else if (*j == ' ' && *(j + 1) == '+') {
+        end = j;
+        break;
+      }
+    }
+    bool tryAddr2Line = false;
+#else
+    for (char* j = stackStrings[i]; j && *j; ++j) {
+      if (*j == '(') {
+        begin = j;
+      } else if (*j == '+') {
+        end = j;
+        break;
+      }
+    }
+    bool tryAddr2Line = true;
+#endif
+    if (begin && end) {
+      *begin++ = '\0';
+      *end = '\0';
+      // found our mangled name, now in [begin, end)
+
+      int status;
+      char* ret = abi::__cxa_demangle(begin, function, &sz, &status);
+      if (ret) {
+        // return value may be a realloc() of the input
+        function = ret;
+        dprintf(fd, "    %s: %s\n", stackStrings[i], function);
+        tryAddr2Line = false;
+      }
+    }
+    if (tryAddr2Line) {
+      // didn't find the mangled name, just print the whole line
+      dprintf(fd, "    %s: ", stackStrings[i]);
+      if (stackTrace[i] && hasExe) {
+        char syscom[4096 + PATH_MAX];
+
+        // Find c++filt from the environment
+        // This is needed for platforms where we still need c++filt -r
+        char const* cxxfilt = getenv("CXXFILT");
+        if (cxxfilt == nullptr) {
+          cxxfilt = "c++filt";
+        }
+        // Do the same for addr2line, just in case we wanted to pass some options
+        char const* addr2line = getenv("ADDR2LINE");
+        if (addr2line == nullptr) {
+          addr2line = "addr2line";
+        }
+        snprintf(syscom, 4096, "%s %p -p -s -f -e %.*s 2>/dev/null | %s ", addr2line, stackTrace[i], exeSize, exe, cxxfilt); // last parameter is the name of this app
+
+        FILE* fp;
+        char path[1024];
+
+        fp = popen(syscom, "r");
+        if (fp == nullptr) {
+          dprintf(fd, "-- no source could be retrieved --\n");
+          continue;
+        }
+
+        while (fgets(path, sizeof(path) - 1, fp) != nullptr) {
+          dprintf(fd, "    %s", path);
+        }
+
+        pclose(fp);
+      } else {
+        dprintf(fd, "-- no source avaliable --\n");
+      }
+    }
+    free(function);
+  }
+  free(stackStrings); // malloc()ed by backtrace_symbols
+  fsync(fd);
+}
+} // namespace o2::framework

--- a/Framework/Foundation/src/RuntimeError.cxx
+++ b/Framework/Foundation/src/RuntimeError.cxx
@@ -15,11 +15,11 @@
 #include <climits>
 #include <atomic>
 #include <cstdarg>
-#include <execinfo.h>
+#include <cstring>
 #include <unistd.h>
 #include <cstdlib>
-#include <cstring>
 #include <cxxabi.h>
+#include <execinfo.h>
 #include <stdexcept>
 
 namespace o2::framework
@@ -71,7 +71,7 @@ RuntimeErrorRef runtime_error_f(const char* format, ...)
   va_start(args, format);
   vsnprintf(gError[i].what, RuntimeError::MAX_RUNTIME_ERROR_SIZE, format, args);
   va_end(args);
-  gError[i].maxBacktrace = canDumpBacktrace() ? backtrace(gError[i].backtrace, RuntimeError::MAX_BACKTRACE_SIZE) : 0;
+  gError[i].maxBacktrace = canDumpBacktrace() ? backtrace(gError[i].backtrace, BacktraceHelpers::MAX_BACKTRACE_SIZE) : 0;
   return RuntimeErrorRef{i};
 }
 
@@ -83,116 +83,13 @@ RuntimeErrorRef runtime_error(const char* s)
     ++i;
   }
   strncpy(gError[i].what, s, RuntimeError::MAX_RUNTIME_ERROR_SIZE);
-  gError[i].maxBacktrace = canDumpBacktrace() ? backtrace(gError[i].backtrace, RuntimeError::MAX_BACKTRACE_SIZE) : 0;
+  gError[i].maxBacktrace = canDumpBacktrace() ? backtrace(gError[i].backtrace, BacktraceHelpers::MAX_BACKTRACE_SIZE) : 0;
   return RuntimeErrorRef{i};
 }
 
 void throw_error(RuntimeErrorRef ref)
 {
   throw ref;
-}
-
-void demangled_backtrace_symbols(void** stackTrace, unsigned int stackDepth, int fd)
-{
-  char** stackStrings;
-  stackStrings = backtrace_symbols(stackTrace, stackDepth);
-  char exe[PATH_MAX];
-  bool hasExe = false;
-  int exeSize = 0;
-
-#if __linux__
-  exeSize = readlink("/proc/self/exe", exe, PATH_MAX);
-  if (exeSize == -1) {
-    dprintf(fd, "Unable to detect exectuable name\n");
-    hasExe = false;
-  } else {
-    dprintf(fd, "Executable is %.*s\n", exeSize, exe);
-    hasExe = true;
-  }
-#endif
-
-  for (size_t i = 1; i < stackDepth; i++) {
-
-    size_t sz = 64000; // 64K ought to be enough for our templates...
-    char* function = static_cast<char*>(malloc(sz));
-    char *begin = nullptr, *end = nullptr;
-    // find the last space and address offset surrounding the mangled name
-#if __APPLE__
-    for (char* j = stackStrings[i]; *j; ++j) {
-      if (*j == ' ' && *(j + 1) != '+') {
-        begin = j;
-      } else if (*j == ' ' && *(j + 1) == '+') {
-        end = j;
-        break;
-      }
-    }
-    bool tryAddr2Line = false;
-#else
-    for (char* j = stackStrings[i]; j && *j; ++j) {
-      if (*j == '(') {
-        begin = j;
-      } else if (*j == '+') {
-        end = j;
-        break;
-      }
-    }
-    bool tryAddr2Line = true;
-#endif
-    if (begin && end) {
-      *begin++ = '\0';
-      *end = '\0';
-      // found our mangled name, now in [begin, end)
-
-      int status;
-      char* ret = abi::__cxa_demangle(begin, function, &sz, &status);
-      if (ret) {
-        // return value may be a realloc() of the input
-        function = ret;
-        dprintf(fd, "    %s: %s\n", stackStrings[i], function);
-        tryAddr2Line = false;
-      }
-    }
-    if (tryAddr2Line) {
-      // didn't find the mangled name, just print the whole line
-      dprintf(fd, "    %s: ", stackStrings[i]);
-      if (stackTrace[i] && hasExe) {
-        char syscom[4096 + PATH_MAX];
-
-        // Find c++filt from the environment
-        // This is needed for platforms where we still need c++filt -r
-        char const* cxxfilt = getenv("CXXFILT");
-        if (cxxfilt == nullptr) {
-          cxxfilt = "c++filt";
-        }
-        // Do the same for addr2line, just in case we wanted to pass some options
-        char const* addr2line = getenv("ADDR2LINE");
-        if (addr2line == nullptr) {
-          addr2line = "addr2line";
-        }
-        snprintf(syscom, 4096, "%s %p -p -s -f -e %.*s 2>/dev/null | %s ", addr2line, stackTrace[i], exeSize, exe, cxxfilt); // last parameter is the name of this app
-
-        FILE* fp;
-        char path[1024];
-
-        fp = popen(syscom, "r");
-        if (fp == nullptr) {
-          dprintf(fd, "-- no source could be retrieved --\n");
-          continue;
-        }
-
-        while (fgets(path, sizeof(path) - 1, fp) != nullptr) {
-          dprintf(fd, "    %s", path);
-        }
-
-        pclose(fp);
-      } else {
-        dprintf(fd, "-- no source avaliable --\n");
-      }
-    }
-    free(function);
-  }
-  free(stackStrings); // malloc()ed by backtrace_symbols
-  fsync(fd);
 }
 
 } // namespace o2::framework

--- a/Framework/Foundation/test/test_RuntimeError.cxx
+++ b/Framework/Foundation/test/test_RuntimeError.cxx
@@ -11,6 +11,7 @@
 
 #include <catch_amalgamated.hpp>
 #include "Framework/RuntimeError.h"
+#include "Framework/BacktraceHelpers.h"
 #include <unistd.h>
 #include <execinfo.h>
 
@@ -33,7 +34,7 @@ TEST_CASE("TestRuntimeError")
     REQUIRE(strncmp(err.what, "foo", 3) == 0);
 #ifdef DPL_ENABLE_BACKTRACE
     backtrace_symbols_fd(err.backtrace, err.maxBacktrace, STDERR_FILENO);
-    o2::framework::demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
+    o2::framework::BacktraceHelpers::demangled_backtrace_symbols(err.backtrace, err.maxBacktrace, STDERR_FILENO);
 #endif
   }
 }

